### PR TITLE
Added type definitions for react-hot-loader

### PR DESCRIPTION
**Pull request by jacekjagiello**
Please fill in this template.

- [x] Prefer to make your PR against the `types-2.0` branch.
- [x] Test the change in your own code.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes).

If adding a new definition:
- [x] The package does not provide its own types, and you can not add them.
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Run `tsc` without errors.
- [x] Include the required [files](https://github.com/DefinitelyTyped/DefinitelyTyped#create-a-new-package) and header. Base these on the README, *not* on an existing project.

If changing an existing definition:
- [] Provide a URL to  documentation or source code which provides context for the suggested changes: 
- [] Increase the version number in the header if appropriate.

### CODE REVIEW

File: react-hot-loader/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+declare module "react-hot-loader" {
+
+  interface ErrorReporterProps {
+    error: any
+  }
+
+  export interface AppContainerProps {
+    children?: React.ReactElement<any>,
+    errorReporter?: React.StatelessComponent<ErrorReporterProps>
```
Comment (Kovensky) [inline react-hot-loader/index.d.ts]: The `errorReporter` type can also be a `React.ComponentClass`, or basically anything you can give as the first argument to `React.createElement` besides `string`.

Though it appears to typecheck even with a class component?

File: react-hot-loader/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+import * as React from "react"
+
+declare module "react-hot-loader" {
+
+  interface ErrorReporterProps {
+    error: any
+  }
+
+  export interface AppContainerProps {
+    children?: React.ReactElement<any>,
```
Comment (Kovensky) [inline react-hot-loader/index.d.ts]: `children` is not optional (the impl will throw if it doesn't get exactly one child)

File: react-hot-loader/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
+// Type definitions for react-hot-loader
+// Project: https://github.com/gaearon/react-hot-loader
+// Definitions by: Jacek Jagiello <https://github.com/jacekjagiello/>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
```
Comment (Kovensky) [inline react-hot-loader/index.d.ts]: Is there any way to note that this applies to version 3 of `react-hot-loader`?

Version 1 has a completely different API (and a lot of disadvantages, such as only supporting createClass components).

File: react-hot-loader/index.d.ts (hunk 2 above)
Comment (JacekJagiello) [inline react-hot-loader/index.d.ts]: You are right, but there is a problem with stateless components and `children` props, which probably will be resolved in typescript 2.1: https://github.com/Microsoft/TypeScript/issues/8588

File: react-hot-loader/index.d.ts (hunk 1 above)
Comment (JacekJagiello) [inline react-hot-loader/index.d.ts]: If a change it to `React.ComponentClass`, then stateless component don't work: 

`Type '({error}: ErrorReporterProps) => Element' is not assignable to type 'ReactElement<ErrorReporterProps>'`

But
`React.ComponentClass<ErrorReporterProps> | React.ComponentClass<ErrorReporterProps>`
seems to work. I'am typescript beginner, do you have any better idea?

File: react-hot-loader/index.d.ts (hunk 3 above)
Comment (JacekJagiello) [inline react-hot-loader/index.d.ts]: Actually, this definition is made for version 3(3.0.0-beta.6), I will add version number.

**Comment by JacekJagiello** [comment]
Seems like there is Travis time-out:
`The job exceeded the maximum time limit for jobs, and has been terminated.`
What should we do with it?

**Comment by andy-ms** [comment]
@JacekJagiello Please merge changes from `types-2.0` to fix tests.

**Comment by JacekJagiello** [comment]
All done, can we merge it?

File: react-hot-loader/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
+// Type definitions for react-hot-loader 3.0.0-beta.6
+// Project: https://github.com/gaearon/react-hot-loader
+// Definitions by: Jacek Jagiello <https://github.com/jacekjagiello/>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+import * as React from "react"
+
+declare module "react-hot-loader" {
```
Comment (andy-ms) [inline react-hot-loader/index.d.ts]: Should be safe to remove this wrapper and just use the file itself as the module definition.

File: react-hot-loader/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,20 @@
+// Type definitions for react-hot-loader 3.0.0-beta.6
```
Comment (andy-ms) [inline react-hot-loader/index.d.ts]: We ignore patch version, so just use `3.0`.

File: react-hot-loader/tsconfig.json [hunk 6]
```diff
@@ -0,0 +1,20 @@
+{
+    "compilerOptions": {
+        "module": "commonjs",
+        "target": "es6",
+        "noImplicitAny": true,
+        "strictNullChecks": false,
```
Comment (andy-ms) [inline react-hot-loader/tsconfig.json]: Make this `true`.

**Comment by JacekJagiello** [comment]
@andy-ms Done, thanks for review

**Pull request merged**